orm/serial_bucket.go

```diff
@@ -0,0 +1,389 @@
+package orm
+
+import (
+	"reflect"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+)
+
+// SerialModel is implemented by any entity that can be stored using SerialModelBucket.
+// GetID/SetID are used to store and access the Key.
+// The ID is always set to nil before serializing and storing the Value.
+type SerialModel interface {
+	weave.Persistent
+	Validate() error
+	GetID() []byte
+	SetID([]byte) error
+}
+
+// SerialModelSlicePtr represents a pointer to a slice of SerialModels. Think of it as
+// *[]SerialModel Because of Go type system, using []SerialModel would not work for us.
+// Instead we use a placeholder type and the validation is done during the
+// runtime.
+type SerialModelSlicePtr interface{}
+
+// SerialModelBucket is implemented by buckets that operates on SerialModels rather than
+// Objects.
+type SerialModelBucket interface {
+	// GetByID query the database for a single SerialModel instance by ID. Lookup is done
+	// by the primary index key. Result is loaded into given destination
+	// SerialModel.
+	// This method returns ErrNotFound if the entity does not exist in the
+	// database.
+	// If given SerialModel type cannot be used to contain stored entity, ErrType
+	// is returned.
+	GetByID(db weave.ReadOnlyKVStore, key []byte, dest SerialModel) error
+
+	// PrefixScan will scan for all SerialModels with a primary key (ID)
+	// that begins with the given prefix.
+	// The function returns a (possibly empty) iterator, which can
+	// load each SerialModel as it arrives.
+	// If reverse is true, iterates in descending order (highest value first),
+	// otherwise, it iterates in ascending order.
+	PrefixScan(db weave.ReadOnlyKVStore, prefix []byte, reverse bool) (SerialModelIterator, error)
+
+	// ByIndex returns all objects that secondary index with given name and
+	// given key. Main index is always unique but secondary indexes can
+	// return more than one value for the same key.
+	// All matching entities are appended to given destination slice. If no
+	// result was found, no error is returned and destination slice is not
+	// modified.
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest SerialModelSlicePtr) error
```

Comment:
I like the iterator concept overall. It abstracts this logic into a basic method:

```go
slice := reflect.ValueOf(dest).Elem()
for i := 0; i < slice.Len(); i++ {
	item := slice.Index(i)
}
```

Plus, if you take a look at `ByIndex` methods underlying workings it's obvious that there is a `scan` and iterated job is going on. Logic could be wrapped with an `Iterator` seamlessly

Reply:
But also there is the case where users would like the ability to load all objects to the memory with an index so that they can access index + 3rd object? In order to achieve both, we can create a new method with a name like `ByIndexLazy` that returns `SerialModelIterator`. But the necessity of the method should be determined by you as the frequent users and developers of **Weave**. Do you think what do you think?

Reply:
This is actually one point where I see the use of a global helper function.
I think best the queries that can return multiple values all return **lazy** iterators. This is especially important to allow Limiting the number of responses for public api. (So I can't query `/accounts?prefix` with `""` data and list all accounts, which could be very heavy for the node.).

Some nice helper functions on the orm package: `WithLimit` and `ToSlice`:

```go
// consumes iter until 50 items, then ends early
lim := orm.WithLimit(iter, 50)

// consume all items in the iterator
var data []MyModel
err := orm.ToSlice(lim, &data)
```

In this model, `orm.ToSlice(bucket.IndexScan(db, index, prefix, reverse), &data)` would do what `bucket.ByIndex(db, index, prefix, reverse, &data)` does now, but in a more composible way (eg. easy to add the limit filter there without overwhelming the argument list)

Reply:
I think @ethanfrey understands my suggestion well.

Having all functions that can return more than one model result return an iterator is in my opinion a nice unification. With proposed by Ethan helpers (`orm.WithLimit` and `orm.ToSlice`), we provide the necessary tooling to work with the iterators. 

This approach reminds me of `io.Reader` interface and [`LimitReader`](https://golang.org/pkg/io/#LimitedReader) and [`ReadAll`](https://golang.org/pkg/io/ioutil/#ReadAll) functions.